src/lib/shadow-shim-helper-rs/src/rootedcell/rc.rs

```diff
@@ -261,3 +279,171 @@ mod test_rooted_rc {
         rc.safely_drop(&root.lock().unwrap());
     }
 }
+
+pub struct RootedRcWeak<T> {
+    tag: Tag,
+    // Option<NonNull<_>> here instead of `* mut` for
+    // [covariance](https://doc.rust-lang.org/reference/subtyping.html#variance).
+    internal: Option<NonNull<RootedRcInternal<T>>>,
+}
+
+impl<T> RootedRcWeak<T> {
+    // Validates that no other thread currently has access to self.internal, and
+    // return a reference to it.
+    fn validate_thread_exclusive_access(&self, root: &Root) -> &RootedRcInternal<T> {
+        assert_eq!(
+            root.tag, self.tag,
+            "Tried using root {:?} instead of {:?}",
+            root.tag, self.tag
+        );
+        // SAFETY:
+        // * Holding a reference to `root` proves no other threads can currently
+        //   access `self.internal`.
+        // * `self.internal` is accessible since we hold a reference.
+        unsafe { self.internal.unwrap().as_ref() }
+    }
+
+    pub fn downgrade(root: &Root, x: &RootedRc<T>) -> Self {
+        x.validate_thread_exclusive_access(root).inc_weak();
+        Self {
+            tag: x.tag,
+            internal: x.internal,
+        }
+    }
+
+    pub fn upgrade(&self, root: &Root) -> Option<RootedRc<T>> {
+        let internal = self.validate_thread_exclusive_access(root);
+
+        if internal.strong_count.get() == 0 {
+            return None;
+        }
+
+        internal.inc_strong();
+
+        Some(RootedRc {
+            tag: self.tag,
+            internal: self.internal,
+        })
+    }
+
+    /// Like [Clone::clone], but requires that the corresponding Root is held.
+    ///
+    /// Intentionally named clone to shadow Self::deref()::clone().
+    ///
+    /// Panics if `root` is not the associated [Root].
+    pub fn clone(&self, root: &Root) -> Self {
+        self.validate_thread_exclusive_access(root).inc_weak();
+        Self {
+            tag: self.tag,
+            internal: self.internal,
+        }
+    }
+
+    pub fn safely_drop(mut self, root: &Root) {
+        let internal = self.validate_thread_exclusive_access(root);
+        internal.dec_weak();
+        let strong_count = internal.strong_count.get();
+        let weak_count = internal.weak_count.get();
+        drop(internal);
```

Comment:
Add a comment about why this `drop` is needed (I'm assuming because the `RootedRcInternal` is freed below)?

Reply:
Unneeded; removed